uniffi_macros/src/export/scaffolding.rs

```diff
@@ -188,25 +188,10 @@ fn gen_ffi_function(
         quote! {
             #[doc(hidden)]
             #[no_mangle]
-            pub extern "C" fn #ffi_ident(
-                #(#params,)*
-                uniffi_executor_handle: ::uniffi::ForeignExecutorHandle,
-                uniffi_callback: <#return_ty as ::uniffi::FfiConverter<crate::UniFfiTag>>::FutureCallback,
-                uniffi_callback_data: *const (),
-                uniffi_call_status: &mut ::uniffi::RustCallStatus,
-            ) {
+            pub extern "C" fn #ffi_ident(#(#params,)*) -> ::uniffi::RustFutureHandle {
                 ::uniffi::deps::log::debug!(#name);
-                ::uniffi::rust_call(uniffi_call_status, || {
-                    #pre_fn_call;
-                    let uniffi_rust_future = ::uniffi::RustFuture::<_, #return_ty, crate::UniFfiTag>::new(
-                        #future_expr,
-                        uniffi_executor_handle,
-                        uniffi_callback,
-                        uniffi_callback_data
-                    );
-                    uniffi_rust_future.wake();
-                    Ok(())
-                });
+                #pre_fn_call;
+                ::uniffi::rust_future_new(async move { #future_expr.await }, crate::UniFfiTag)
```

Comment:
Is new `async move {}` actually fixes an important bug.  If you have a [async method that inputs `&self`](https://github.com/bendk/uniffi-rs/blob/3b7180f374fcd27f2abc66c8f3c1c3787ee57e04/fixtures/metadata/src/tests.rs#L384-L386), then the generate future has the same lifetime as `self`.  Things would break if you destroyed the object then continued to poll the future, which was possible with the old system.

I found this out because Rust complained when compiling the `fixtures/metadata` test.  In general, it seems like the code in `rustfutures.rs` relies less on unsafe code and our intuition and therefore Rust is able to check things better.  It was really annoying adding all those `Send + 'static` bounds, but that's already paid off.